### PR TITLE
Update RAFT documentation

### PR DESCRIPTION
**Pull request by lowener**
- Various documentation updates on C++ and Python doc, mainly for raft::neighbors
- Add QPS vs Recall plot

### CODE REVIEW

**Review by cjnolet — COMMENTED** [review]
Overall I think the changes look great. I'm holding off on approving only because of the format changes for the code examples in the python docs. We had originally shifted these examples to a different format and I don't know if we want to revert them. 

cc @benfred @galipremsagar for their thoughts here.

File: python/pylibraft/pylibraft/cluster/kmeans.pyx
Comment (galipremsagar) [inline python/pylibraft/pylibraft/cluster/kmeans.pyx:87]: @lowener Any reason why we are adding python code-blocks? AFAIK sphinx does a good job at parsing the example code section of documentation.

Comment (lowener) [reply python/pylibraft/pylibraft/cluster/kmeans.pyx:87]: Currently the code examples on Sphinx are split into many small blocks, I thought that it was unintentional and I didn't know that it was a shift to a different format.
If this small block format has already been decided then I should revert my changes.
Current
image
Proposed
image

File: python/pylibraft/pylibraft/cluster/kmeans.pyx
Comment (galipremsagar) [inline python/pylibraft/pylibraft/cluster/kmeans.pyx:90]:
```suggestion
```
@lowener To slove the issue you are seeing in the docs, you just have to delete these empty lines instead of creating a python code-block.. Good catch!

File: python/pylibraft/pylibraft/cluster/kmeans.pyx
Comment (galipremsagar) [reply python/pylibraft/pylibraft/cluster/kmeans.pyx:87]: Got it, I have a simpler solution for it. See my comment below: https://github.com/rapidsai/raft/pull/1717/files#r1284444539

File: python/pylibraft/pylibraft/cluster/kmeans.pyx
Comment (lowener) [reply python/pylibraft/pylibraft/cluster/kmeans.pyx:90]: One last argument that I have in favor of the code-block is that it improves readability. In the previous example there is small breaks between the includes, the variable declaration, and the RAFT code. I think it can be appreciated by our Python users. What do you think?

Comment (cjnolet) [reply python/pylibraft/pylibraft/cluster/kmeans.pyx:90]: IIUC, the .. code block is a sphinx-specific documentation style while the other format is more universal across python. Is that true?

Comment (lowener) [reply python/pylibraft/pylibraft/cluster/kmeans.pyx:90]: From what I understand as well yes. When running `?cagra.build` in IPython for details about this function users would see written literally `.. code-block::python`.

Comment (galipremsagar) [reply python/pylibraft/pylibraft/cluster/kmeans.pyx:90]: Yeah, that's right. Lets drop `code-block` usages in that case.

**Review by galipremsagar — APPROVED** [review]

**Review by cjnolet — APPROVED** [review]
Changes LGTM! If we are choosing to document simple param structs at the class level, we should be as consistent as possible and make sure we do it everywhere. 

Looks like the doctest pytests are failing. We should be able to merge after we fix those.

**Pull request merged**
